ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -246,3 +231,31 @@ let disable_clustering ~__context =
      info "Disabling cluster_host %s" (Ref.string_of self);
      disable ~__context ~self
 
+let sync_required ~__context ~host =
+  let clusters = Db.Cluster.get_all_records ~__context in
+  match clusters with
+  | [] -> None
+  | [cluster_ref, cluster_rec] -> begin
+      let expr = Db_filter_types.(And (Eq (Field "host", Literal (Ref.string_of host)),
+                                       Eq (Field "cluster", Literal (Ref.string_of cluster_ref)))) in
+      let my_cluster_hosts = Db.Cluster_host.get_internal_records_where ~__context ~expr in
+      match my_cluster_hosts with
+      | [(_ref,_rec)] -> None
+      | [] ->
+        if cluster_rec.API.cluster_pool_auto_join
+        then Some cluster_ref
+        else None
+      | _ -> raise Api_errors.(Server_error (internal_error, [ "Host cannot be associated with more than one cluster_host"; Ref.string_of host ]))
+    end
+  | _ -> raise Api_errors.(Server_error (internal_error, ["Cannot have more than one Cluster object per pool currently"]))
+
+(* If cluster found without local cluster_host, create one in db *)
+let create_as_necessary ~__context ~host =
+  match sync_required ~__context ~host with
+  | Some cluster -> (* assume pool autojoin set *)
+    let network = get_network_internal ~__context ~self:cluster in
+    let (pIF,pifrec) = Xapi_clustering.pif_of_host ~__context network host in
+    fix_pif_prerequisites ~__context (pIF,pifrec);
```

Comment:
This is in the wrong place, it'll fail: we have no IP on startup where we create the object, this should be part of join.